bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/evaluation/EvaluationInteractor.kt

```diff
@@ -0,0 +1,83 @@
+package io.bucketeer.sdk.android.internal.evaluation
+
+import android.annotation.SuppressLint
+import android.content.SharedPreferences
+import androidx.annotation.VisibleForTesting
+import io.bucketeer.sdk.android.internal.Constants
+import io.bucketeer.sdk.android.internal.evaluation.db.EvaluationDao
+import io.bucketeer.sdk.android.internal.logd
+import io.bucketeer.sdk.android.internal.loge
+import io.bucketeer.sdk.android.internal.model.Evaluation
+import io.bucketeer.sdk.android.internal.model.User
+import io.bucketeer.sdk.android.internal.remote.ApiClient
+import io.bucketeer.sdk.android.internal.remote.GetEvaluationsResult
+
+/**
+ * Evaluation business logics.
+ *
+ * All methods in this class must be sync to keep things simple.
+ */
+internal class EvaluationInteractor(
+  private val apiClient: ApiClient,
+  private val evaluationDao: EvaluationDao,
+  private val sharedPrefs: SharedPreferences,
+) {
+  // key: userId
+  @VisibleForTesting
+  internal val evaluations = mutableMapOf<String, List<Evaluation>>()
+
+  @VisibleForTesting
+  internal var currentEvaluationsId: String
+    get() = sharedPrefs.getString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, "") ?: ""
+    @SuppressLint("ApplySharedPref")
+    set(value) {
+      sharedPrefs.edit()
+        .putString(Constants.PREFERENCE_KEY_USER_EVALUATION_ID, value)
+        .commit()
+    }
+
+  @Suppress("MoveVariableDeclarationIntoWhen")
+  fun fetch(user: User, timeoutMillis: Long?): GetEvaluationsResult {
+    val currentEvaluationsId = this.currentEvaluationsId
+
+    val result = apiClient.getEvaluations(user, currentEvaluationsId, timeoutMillis)
+
+    when (result) {
+      is GetEvaluationsResult.Success -> {
+        val response = result.value.data
+        val newEvaluationsId = response.user_evaluations_id
+        if (currentEvaluationsId == newEvaluationsId) {
+          logd { "Nothing to sync" }
+          return result
+        }
+
+        val newEvaluations = response.evaluations.evaluations
+
+        val success = evaluationDao.deleteAllAndInsert(user.id, newEvaluations)
+        if (!success) {
+          loge { "Failed to update latest evaluations" }
+          return result
+        }
+
+        this.currentEvaluationsId = newEvaluationsId
+
+        evaluations[user.id] = newEvaluations
+      }
+      is GetEvaluationsResult.Failure -> {
+        logd(result.error) { "ApiError: ${result.error.message}" }
+      }
+    }
+    return result
+  }
+
+  fun refreshCache() {
```

Comment:
Simple refresh method.
This is going to be called upon start-up time.

Reply:
Just confirming, is it ok to use `groupBy` for this case?

Reply:
do you have any concerns?

Reply:
I think it's ok as long as Evaluation.user_id has value

Reply:
I asked because I would use the `WHERE` clause instead of `GROUP BY`.
But as you said as long the user_id column has a value. It's ok.
Also, it won't have too much data, so there is no performance issue.
Thanks!

Reply:
oh I see.

I thought about that too, but in theory, the table could have multiple user's data, and I'm not sure if we should load all the data or just current user's data.

But　as far as I can see from BKTClient interface, user is not swappable.
I'm going to use `WHERE` clause as you suggested.
thanks!

Reply:
I'll open a separate PR for this.